### PR TITLE
Code cleanup on Membership form when freezing fields in a recurring exists situation

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Minor code extraction & tidy up

Before
----------------------------------------
an array of $elements is built as fields are added. It takes work to figure out why

After
----------------------------------------
Elements to be frozen are frozen as they are added.

Technical Details
----------------------------------------
I actually was trying to extract out the handling of the date fields to the addEntityFieldsToTemplate() function & convert them to datepicker but I wound up feeling I had to unravel this first

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Member/Form/Membership.php [hunk 1]
```diff
@@ -1940,4 +1928,19 @@ protected function setStatusMessage($membership, $endDate, $receiptSend, $member
...
 
+  /**
+   * @return bool
+   */
+  protected function isUpdateToExistingRecurringMembership() {
+    $isRecur = FALSE;
+    if ($this->_action & CRM_Core_Action::UPDATE
+      && CRM_Core_DAO::getFieldValue('CRM_Member_DAO_Membership', $this->_id,
+        'contribution_recur_id')
+      && !CRM_Member_BAO_Membership::isSubscriptionCancelled($this->_id)) {
```
Comment (mattwire) [inline CRM/Member/Form/Membership.php]: Is this function coming out of some existing code?  I *think* it makes sense but I need to be sure it's not a change in behaviour? - ie. this is assuming that a cancelled membership cannot be renewed and that may not previously have been the case (even if it's the right thing to do)

**Comment by mattwire** [comment]
@eileenmcnaughton Could do with a bit of context here - is this a step towards merging Membership and MembershipRenewal forms? Is the freezing elements new for the Membership form, but existing on the MembershipRenewal forms?

File: CRM/Member/Form/Membership.php (hunk 1 above)
Comment (seamuslee001) [reply CRM/Member/Form/Membership.php]: @mattwire check old lines 666 - 671

**Comment by eileenmcnaughton** [comment]
@mattwire yeah  - a very preliminary step IMHO. When I looked at Membership form the 2 steps I thought should be first were

1) replacing the jcalendar fields 
2) setting the status message 'as we go' rather than building $createdMemberships & passing to     ```$this->setStatusMessage($membership, $endDate, $receiptSent, $membershipTypes, $createdMemberships, $isRecur, $calcDates, $mailSend);``` - This latter is because in order to rationalise post process code it is necessary to reduct / sanitise what variables are needed in later parts of the code.

When I went to do 1 I added the entityFieldsTrait & defined the 3 fields in the setEntityFields & changed the tpl - I got it close to working but realised the setEntityFields needed to be able to specify whether end_date if freezable.

**Comment by mattwire** [comment]
* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) Pass
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) Pass
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) Pass
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) Pass

Good to merge @eileenmcnaughton @seamuslee001

**Comment by seamuslee001** [comment]
Merging as per @mattwire's review

**Comment by eileenmcnaughton** [comment]
thanks @mattwire

**Pull request merged**